src/sciline/reporter.py

```diff
@@ -0,0 +1,256 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+import functools
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Hashable, Iterable
+from types import TracebackType
+from typing import Any
+
+from ._provider import Provider
+from ._utils import provider_name
+
+
+class Reporter(ABC):
+    """Base class for progress reporters of computations."""
+
+    def __init__(self) -> None:
+        self._n_steps = 0
+        self._provider_id = -1
+
+    def run_computation(self, providers: Iterable[Provider]) -> Reporter:
+        """Run a single computation with progress reporting.
+
+        A ``Reporter`` instance must not be used for multiple computations concurrently!
+
+        Use this method in a ``with`` statement:
+
+        .. code-block:: python
+
+            with reporter.run_computation(providers):
+                # Perform computation
+
+        This method exists mainly to provide arguments when entering a reporting
+        context, implementers should mainly override ``__enter__``.
+        """
+        self._n_steps = sum(1 for provider in providers if provider.kind == "function")
```

Comment:
This trips up the case when normal functions are used to create the pipeline. The example at https://scipp.github.io/sciline/#use fails with `AttributeError: 'function' object has no attribute 'kind'` as the task graph isn't created which handles the conversion to `Provider` type.
Or have got the usage wrong here? I tried with:
```python
reporter = sciline.reporter.RichReporter()
with reporter.run_computation(providers=providers):
... 
```

Reply:
You should not use the context manager yourself. Instead use (adapted from your link)
```python
reporter = sciline.reporter.RichReporter()
providers = [load, clean, process]
params = {Filename: 'data.txt', ScaleFactor: 2.0}
pipeline = sciline.Pipeline(providers, params=params)

pipeline.compute(Result, reporter=reporter)
```

We should never pass `Provider` objects to a pipeline. They are created internally.